src/v/kafka/client/direct_consumer/direct_consumer.cc

```diff
@@ -0,0 +1,186 @@
+#include "kafka/client/direct_consumer/direct_consumer.h"
+
+#include "kafka/client/direct_consumer/data_queue.h"
+#include "kafka/client/direct_consumer/fetcher.h"
+#include "model/validation.h"
+
+namespace kafka::client {
+
+direct_consumer::direct_consumer(cluster& cluster, configuration cfg)
+  : _cluster(&cluster)
+  , _config(std::move(cfg))
+  , _fetched_data_queue(std::make_unique<data_queue>())
+  , _metadata_callback_id(_cluster->register_metadata_cb(
+      [this](const metadata_response_data& d) { on_metadata_update(d); })) {}
+
+ss::future<fetches>
+direct_consumer::fetch_next(std::chrono::milliseconds timeout) {
+    auto holder = _gate.hold();
+    try {
+        co_return co_await _fetched_data_queue->pop(timeout);
+    } catch (ss::condition_variable_timed_out&) {
+        co_return chunked_vector<fetched_topic_data>{};
+    }
+}
+
+void direct_consumer::update_configuration(configuration cfg) { _config = cfg; }
+
+ss::future<> direct_consumer::update_fetchers() {
+    // Stop fetch loops by acquiring a write lock
+    auto u = co_await _fetchers_lock.hold_write_lock();
+    for (auto& [topic, partitions] : _removals) {
+        for (auto& p : partitions) {
+            for (auto& [id, fetcher] : _broker_fetchers) {
+                // Unassign partitions from all fetchers
+                co_await fetcher->unassign_partition(
+                  model::topic_partition_view(topic, p));
+            }
+        }
+    }
+    for (auto& [topic, partitions] : _subscriptions) {
+        for (auto& [p_id, sub] : partitions) {
+            auto leader_id = _cluster->get_topics().leader(
+              model::topic_partition_view(topic, p_id));
+
+            if (!leader_id) {
+                ssx::spawn_with_gate(_gate, [this] {
+                    return _cluster->request_metadata_update();
```

Comment:
I think _cluster should expose a queue_metadata_update() or something like that?